### PR TITLE
Buffer API

### LINKED ISSUE #1994: Buffer API
As part of https://github.com/xtermjs/xterm.js/issues/1128 we will be pulling the search addon out of the core repo and into a satellite repo to be published as its own npm module. This is exactly the sort of thing that would be cool for addons to be able to build so we should stabilize some API as right now it hooks into internals. Here's what it touches in buffer:

- ydisp
- ybase
- length
- translateBufferLineToString
- lines (for isWrapped, length, get width, get char to check if emoji)

Given this, I'm proposing the following API. It resembles the internal structure, but I tweaked a bit here and there to make it more consistent and easier to use (eg. `buffer.lines.get` -> `buffer.getLine`):

```ts
class Terminal {
    /**
     * The terminal's current buffer, note that this might be either the normal
     * buffer or the alt buffer depending on what's running in the terminal.
     */
    buffer: IBuffer;
}

interface IBuffer {

    /**
     * The y position of the cursor. This ranges between `0` (when the
     * cursor is at baseY) and `Terminal.rows - 1` (when the cursor is on the
     * last row).
     */
    readonly cursorY: number;

    /**
     * The x position of the cursor. This ranges between `0` (left side) and
     * `Terminal.cols - 1` (right side).
     */
    readonly cursorX: number;

    /**
     * The line within the buffer where the top of the viewport is.
     */
    readonly viewportY: number;

    /**
     * The line within the buffer where the top of the bottom page is (when
     * fully scrolled down);
     */
    readonly baseY: number;

    /**
     * The amount of lines in the buffer.
     */
    readonly length: number;

    /**
     * Gets a line from the buffer.
     * 
     * @param y The line index to get.
     */
    getLine(y: number): IBufferLine;
}

interface IBufferLine {
    /**
     * Whether the line is wrapped from the previous line.
     */
    readonly isWrapped: boolean;

    /**
     * Gets a cell from the line.
     * 
     * @param x The character index to get.
     */
    getCell(x: number): IBufferCell;

    /**
     * Gets the line
     */
    translateToString(trimRight: boolean, startColumn: number, endColumn: number): string;
}

interface IBufferCell {
    /**
     * The character within the cell.
     */
    readonly char: number;

    /**
     * The width of the character. Some examples:
     * 
     * - This is `1` for most cells.
     * - This is `2` for wide character like CJK glyphs.
     * - This is `0` for cells immediately following cells with a width of `2`.
     */
    readonly width: number;
}
```

### PR DESCRIPTION
**Pull request by Tyriar**
This exposes the primary parts of the buffer to APIs, see changes to xterm.d.ts for the actual public API. It works by having a "view" object in public/ that owns the actual object and exposes friendly aliases for the properties, for example `viewportY` instead of `ydisp`, `cursorX` over `x`, `getLine` over `lines.get(x)`, etc.

The first usage of this will be moving the search addon into the new addon API https://github.com/xtermjs/xterm.js/pull/2065, built 100% on stable API! The buffer API is currently marked experimental to make sure it meets the needs of the search addon, we can remove experimental at or before v4 https://github.com/xtermjs/xterm.js/issues/2075

Fixes #1994

### CODE REVIEW

File: src/public/Terminal.ts [hunk 1]
```diff
@@ -169,3 +171,30 @@ export class Terminal implements ITerminalApi {
...
+
+class BufferApiView implements IBufferApi {
+  constructor(private _buffer: IBuffer) {}
+
+  public get cursorY(): number { return this._buffer.y; }
+  public get cursorX(): number { return this._buffer.x; }
+  public get viewportY(): number { return this._buffer.ydisp; }
+  public get baseY(): number { return this._buffer.ybase; }
+  public get length(): number { return this._buffer.lines.length; }
+  public getLine(y: number): IBufferLineApi { return new BufferLineApiView(this._buffer.lines.get(y)); }
```
Comment (jerch) [inline src/public/Terminal.ts]: Does this need a sanity check to avoid a lazy error due to a non existing line?

File: src/public/Terminal.ts [hunk 2]
```diff
@@ -169,3 +171,30 @@ export class Terminal implements ITerminalApi {
...
+}
+
+class BufferLineApiView implements IBufferLineApi {
+  constructor(private _line: IBufferLine) {}
+
+  public get isWrapped(): boolean { return this._line.isWrapped; }
+  public getCell(x: number): IBufferCellApi { return new BufferCellApiView(this._line, x); }
```
Comment (jerch) [inline src/public/Terminal.ts]: Same here - how about x overflowing the actual length?

**Comment by jerch** [comment]
@Tyriar
The way you currently create the BufferView by referencing is not "stable" for buffer lines between data updates of the terminal due to recycling and line movements. Not sure yet if thats a drawback/issue at all, still it means that every action from an addon will have to re-request the view to see the real terminal buffer state with the correct buffer line order.

I think this is not an issue if we clearly document that ppl should avoid holding refs to members on nonlocal scope (rule of thumb - no event loop in between), instead always refer to them by the full descent like `terminal.buffer().getLine(y).getCell(x)....`. If they really have to hold stuff, they will have to copy the parts over to some own container. Local refs are not affected by this.

**Comment by Tyriar** [comment]
@jerch I think this is fine if it's called out in the documentation as you mention, I'll add some more details to the jsdoc

File: src/public/Terminal.ts (hunk 1 above)
Comment (Tyriar) [reply src/public/Terminal.ts]: This doesn't but it now correctly returns `IBufferLineApi | undefined`

**Pull request merged**